src/misc/cache/lv_cache_entry.c

```diff
@@ -0,0 +1,167 @@
+/**
+* @file lv_cache_entry.c
+*
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_cache_entry.h"
+#include "../../stdlib/lv_sprintf.h"
+#include "../lv_assert.h"
+#include "lv_cache.h"
+#include "lv_cache_entry_private.h"
+#include "lv_cache_private.h"
+/*********************
+ *      DEFINES
+ *********************/
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+struct _lv_cache_entry_t {
+    const lv_cache_t * cache;
+    int32_t ref_cnt;
+    uint32_t node_size;
+
+    bool is_invalid;
+};
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+
+/**********************
+ *  GLOBAL VARIABLES
+ **********************/
+
+/**********************
+ *  STATIC VARIABLES
+ **********************/
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+void lv_cache_entry_reset_ref(lv_cache_entry_t * entry)
+{
+    LV_ASSERT_NULL(entry);
+    entry->ref_cnt = 0;
+}
+void lv_cache_entry_inc_ref(lv_cache_entry_t * entry)
+{
+    LV_ASSERT_NULL(entry);
+    entry->ref_cnt++;
+}
+void lv_cache_entry_dec_ref(lv_cache_entry_t * entry)
+{
+    LV_ASSERT_NULL(entry);
+    entry->ref_cnt--;
+    if(entry->ref_cnt < 0) {
+        LV_LOG_WARN("ref_cnt(%" LV_PRIu32 ") < 0", entry->ref_cnt);
+        entry->ref_cnt = 0;
+    }
+}
+int32_t lv_cache_entry_get_ref(lv_cache_entry_t * entry)
+{
+    LV_ASSERT_NULL(entry);
+    return entry->ref_cnt;
+}
+uint32_t lv_cache_entry_get_node_size(lv_cache_entry_t * entry)
+{
+    return entry->node_size;
+}
+void lv_cache_entry_set_node_size(lv_cache_entry_t * entry, uint32_t node_size)
+{
+    LV_ASSERT_NULL(entry);
+    entry->node_size = node_size;
+}
+void lv_cache_entry_set_invalid(lv_cache_entry_t * entry, bool is_invalid)
+{
+    LV_ASSERT_NULL(entry);
+    entry->is_invalid = is_invalid;
+}
+bool lv_cache_entry_is_invalid(lv_cache_entry_t * entry)
+{
+    LV_ASSERT_NULL(entry);
+    return entry->is_invalid;
+}
+void * lv_cache_entry_get_data(lv_cache_entry_t * entry)
+{
+    LV_ASSERT_NULL(entry);
+    return (uint8_t *)entry - entry->node_size;
```

Comment:
This whole logic is used to simplify users' mallocing a entry and free it, which can be easily done before call custom `free_cache_entry_cb` etc. 

So I suggest malloc entry, and malloc user's data separately. Return entry->data directly from `lv_cache_entry_get_data `, and no need for `lv_cache_entry.c` any more.

Reply:
I don't want the user to be able to manually modify the entry's data, he has to go through an API action

Reply:
I meant adding `void * data` to `lv_cache_entry_t `. User still call `lv_cache_entry_get_data ` to retrieve data.